### PR TITLE
Change ElasticsearchWriter to write perfdata as an array of objects

### LINKED ISSUE #10511: Rework of ElasticsearchWriter performance data
Hi, 

I have some suggestions for improving the ElasticsearchWriter.

Currently the ElasticsearchWriter creates a new field for every metric in a given check command.
This leads to a vast number of fields in the Elasticsearch/OpenSearch indices.

Here are some examples:

```json
{
  "_index": "icinga",
  "_id": "UQdo5ZcByMLqDokf4UFI",
  "_source": {
    "@timestamp": "2025-07-07T15:02:20.573+0000",
    "check_command": "load",
    "check_result.check_source": "homestead",
    "check_result.command": [
      "/usr/lib/nagios/plugins/check_load",
      "-c",
      "10,6,4",
      "-w",
      "5,4,3"
    ],
    "check_result.execution_end": "2025-07-07T15:02:20.573+0000",
    "check_result.execution_start": "2025-07-07T15:02:20.570+0000",
    "check_result.execution_time": 0.0027740001678466797,
    "check_result.exit_status": 0,
    "check_result.latency": 0.0003490447998046875,
    "check_result.output": "OK - load average: 0.24, 0.25, 0.26",
    "check_result.perfdata.load1.crit": 10,
    "check_result.perfdata.load1.min": 0,
    "check_result.perfdata.load1.value": 0.24,
    "check_result.perfdata.load1.warn": 5,
    "check_result.perfdata.load15.crit": 4,
    "check_result.perfdata.load15.min": 0,
    "check_result.perfdata.load15.value": 0.26,
    "check_result.perfdata.load15.warn": 3,
    "check_result.perfdata.load5.crit": 6,
    "check_result.perfdata.load5.min": 0,
    "check_result.perfdata.load5.value": 0.25,
    "check_result.perfdata.load5.warn": 4,
    "check_result.schedule_end": "2025-07-07T15:02:20.573+0000",
    "check_result.schedule_start": "2025-07-07T15:02:20.569+0000",
    "check_result.state": 0,
    "check_result.vars_after": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "check_result.vars_before": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "current_check_attempt": 1,
    "host": "homestead",
    "last_hard_state": 0,
    "last_state": 0,
    "max_check_attempts": 5,
    "reachable": true,
    "service": "load",
    "state": 0,
    "state_type": 1,
    "timestamp": "2025-07-07T15:02:20.573+0000",
    "type": "icinga2.event.checkresult"
  }
},
{
  "_index": "icinga",
  "_id": "-j6O6pcBNgAT8LEo7v53",
  "_score": 1,
  "_source": {
    "@timestamp": "2025-07-08T15:01:56.548+0000",
    "check_command": "ping4",
    "check_result.check_source": "homestead",
    "check_result.command": [
      "/usr/lib/nagios/plugins/check_ping",
      "-4",
      "-H",
      "127.0.0.1",
      "-c",
      "200,15%",
      "-w",
      "100,5%"
    ],
    "check_result.execution_end": "2025-07-08T15:01:56.548+0000",
    "check_result.execution_start": "2025-07-08T15:01:52.466+0000",
    "check_result.execution_time": 4.081341981887817,
    "check_result.exit_status": 0,
    "check_result.latency": 0.0006992816925048828,
    "check_result.output": "PING OK - Packet loss = 0%, RTA = 0.07 ms",
    "check_result.perfdata.pl.crit": 15,
    "check_result.perfdata.pl.min": 0,
    "check_result.perfdata.pl.unit": "percent",
    "check_result.perfdata.pl.value": 0,
    "check_result.perfdata.pl.warn": 5,
    "check_result.perfdata.rta.crit": 0.2,
    "check_result.perfdata.rta.min": 0,
    "check_result.perfdata.rta.unit": "seconds",
    "check_result.perfdata.rta.value": 7.4e-05,
    "check_result.perfdata.rta.warn": 0.1,
    "check_result.schedule_end": "2025-07-08T15:01:56.548+0000",
    "check_result.schedule_start": "2025-07-08T15:01:52.466+0000",
    "check_result.state": 0,
    "check_result.vars_after": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "check_result.vars_before": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "current_check_attempt": 1,
    "host": "Demo1",
    "last_hard_state": 0,
    "last_state": 0,
    "max_check_attempts": 5,
    "reachable": true,
    "service": "ping4",
    "state": 0,
    "state_type": 1,
    "timestamp": "2025-07-08T15:01:56.548+0000",
    "type": "icinga2.event.checkresult"
  }
}
```

This means, the ElasticsearchWriter is practically unusable in large Icinga2 setups. Because a high amount of fields can lead to performance degradations and memory issues in clusters (See https://www.elastic.co/guide/en/elasticsearch/reference/8.18/mapping-settings-limit.html )

Elasticsearch/OpenSearch only allow 1000 fields as default, to safeguard clusters against indices that create too many fields.

As mentioned, the Icinga2 ElasticsearchWriter creates a new field for each metric in a check plugin. Each new plugin, each new metric adds new fields to the index. See also: https://github.com/Icinga/icinga2/issues/6805

# A possible solution

One solution could be to redesign the output of the ElasticsearchWriter to use an array of nested objects for the performance data.
These are supported in both Elasticsearch and OpenSearch:

- https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/nested
- https://docs.opensearch.org/docs/latest/field-types/supported-field-types/nested/

> The nested type is a specialised version of the object data type that allows arrays of objects to be indexed in a way that they can be queried independently of each other.

Example from the docs:

```
{
  "patients": [
    {"name" : "John Doe", "age" : 56, "smoker" : true},
    {"name" : "Mary Major", "age" : 85, "smoker" : false}
  ]
}
```
There are some pitfalls to look out for. By default, arrays of objects are flattened during indexing. Meaning the ElasticsearchWrite needs to make sure the nested field is used.

An example of how this would look like in Icinga2:

```json
{
  "_index": "icinga",
  "_id": "UQdo5ZcByMLqDokf4UFI",
  "_score": 1,
  "_source": {
    "@timestamp": "2025-07-07T15:02:20.573+0000",
    "check_command": "load",
    "check_result.check_source": "homestead",
    "check_result.command": [
      "/usr/lib/nagios/plugins/check_load",
      "-c",
      "10,6,4",
      "-w",
      "5,4,3"
    ],
    "check_result.execution_end": "2025-07-07T15:02:20.573+0000",
    "check_result.execution_start": "2025-07-07T15:02:20.570+0000",
    "check_result.execution_time": 0.0027740001678466797,
    "check_result.exit_status": 0,
    "check_result.latency": 0.0003490447998046875,
    "check_result.output": "OK - load average: 0.24, 0.25, 0.26",
    "check_result.perfdata": [
        {
            "metric_name": "load1",
            "crit": 10,
            "min": 0,
            "value": 0.24,
            "warn": 5,
        },
        {
            "metric_name": "load15",
            "crit": 4,
            "min": 0,
            "value": 0.26,
            "warn": 3,
        },
        {
            "metric_name": "load5",
            "crit": 6,
            "min": 0,
            "value": 0.25,
            "warn": 4,
        },
    ]
    "check_result.schedule_end": "2025-07-07T15:02:20.573+0000",
    "check_result.schedule_start": "2025-07-07T15:02:20.569+0000",
    "check_result.state": 0,
    "check_result.vars_after": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "check_result.vars_before": {
      "attempt": 1,
      "reachable": true,
      "state": 0,
      "state_type": 1
    },
    "current_check_attempt": 1,
    "host": "homestead",
    "last_hard_state": 0,
    "last_state": 0,
    "max_check_attempts": 5,
    "reachable": true,
    "service": "load",
    "state": 0,
    "state_type": 1,
    "timestamp": "2025-07-07T15:02:20.573+0000",
    "type": "icinga2.event.checkresult"
  }
}
```

Benefits:

1. Avoids creating new fields for each metric in a check plugin
2. Make the index fields more predictable when searching
3. Makes the perfdata independently searchable

Obviously, this would be a breaking change. 

However, I don't see how the ElasticsearchWriter would ever be usable practice without reworking the current performance data output, regardless of how it is reworked. Since every cluster is doomed to run into the field mapping limit eventually with the current implementation, as fields cannot be removed from the mapping.

Regards,
Markus

### PR DESCRIPTION
**Pull request by martialblog**
Hi,

Let me preface this PR with: I'm not a C++ developer. So please excuse my C++, tried my best, feedback is welcome.

Prior to this change the ElasticsearchWriter created a new field for every metric in check command. This leads to a vast number of fields in the indices, aka a field mapping explosion.

Now, it uses an array of objects. This avoids the explosion, Makes the index fields more predictable when searching and Makes the perfdata independently searchable.

How the new output looks like:

```json
          "check_result.latency": 0.0007982254028320313,
          "check_result.output": "LOAD OK - total load average: 1.84, 1.77, 2.27",
          "check_result.perfdata": [
            {
              "crit": 10,
              "label": "load1",
              "min": 0,
              "value": 1.84,
              "warn": 5
            },
            {
              "crit": 6,
              "label": "load5",
              "min": 0,
              "value": 1.77,
              "warn": 4
            },
            {
              "crit": 4,
              "label": "load15",
              "min": 0,
              "value": 2.27,
              "warn": 3
            }
          ],
          "check_result.schedule_end": "2025-07-24T10:05:52.877+0000",
          "check_result.schedule_start": "2025-07-24T10:05:52.870+0000",
```

Fixes #10511

### CODE REVIEW

**Comment by martialblog** [comment]
Tested it successfully with:

- elasticsearch:8.18.4
- elasticsearch:9.0.3
- opensearch:2.19.3
- opensearch:3.1.0

**Comment by oxzi** [comment]
First, thanks for your PR.

How is this change reflected in the ElasticSearch? Will this break compatibility due to the format change, especially for custom dashboards?

**Comment by martialblog** [comment]
This will change the format of what is written to Elasticsearch from this:

```json
    "check_result.perfdata.load1.crit": 10,
    "check_result.perfdata.load1.min": 0,
    "check_result.perfdata.load1.value": 0.24,
    "check_result.perfdata.load1.warn": 5,
```

To this:

```json
  "check_result.perfdata": [
 {
  "crit": 10,
  "label": "load1",
  "min": 0,
  "value": 0.24,
  "warn": 5
 }
]
```

Since existing Elasticsearch indices use a dynamic field mapping, no changes will have to be made in Elasticsearch (unless the users have a fixed mapping, but I doubt that, because of the current design of the fields that wouldn't be practical).

But yes, custom dashboards will break.

However, as described in the issue, the current designed was flawed from the beginning. Creating new fields for the metrics is not how this should have been done in the first place. In the SQL world this would be equal to creating a new column for each new metric, which would make no sense.

**Comment by martialblog** [comment]
This issue here also makes some valid points: https://github.com/Icinga/icinga2/issues/6805

> ES can only handle 1000 different fields as default. You can increase the limit, but this will only shift the issue to a later moment. It is also very hard to handle the different perfdata labels in Grafana for example, because I can not just get all perfdata labels for a specific host. With template variables I could query all values with perfdata for a specific host. This is not possible with the current field-solution.

**Comment by oxzi** [comment]
As discussed offline, I generally support this PR, even if it results in breakages. However, I would like to test it before approving it. Since I don't currently have a testing ES setup, this may take a while.

**Comment by martialblog** [comment]
A quick and simple Elasticsearch/OpenSearch setup can be done with:

```
#OS 
podman run --name os2 -p 9201:9200 -m 1g --rm -ti -e "plugins.security.disabled=true" -e "discovery.type=single-node" -e "OPENSEARCH_INITIAL_ADMIN_PASSWORD=Foobar_123456" docker.io/opensearchproject/opensearch:2
podman run --name os3 -p 9201:9200 -m 1g --rm -ti -e "plugins.security.disabled=true" -e "discovery.type=single-node" -e "OPENSEARCH_INITIAL_ADMIN_PASSWORD=Foobar_123456" docker.io/opensearchproject/opensearch:3

# ES
podman run --name es9 -p 9200:9200 -m 1g --rm -it -e "xpack.security.enabled=false" -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:9.0.3
podman run --name es8 -p 9200:9200 -m 1g --rm -it -e "xpack.security.enabled=false" -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:8.18.4
```

Here some example calls:

```
# Get all indices
curl -X GET "http://localhost:9200/_cat/indices"

# Get the mapping of an index
curl -X GET "http://localhost:9200/icinga2/_mapping"

# Get the last docs in an index
curl -X GET "http://localhost:9200/icinga2/_search"
```

**Comment by martialblog** [comment]
@jschmidt-icinga As discussed, I did a rebase on the current master

**Comment by martialblog** [comment]
@jschmidt-icinga Been thinking, maybe we could introduce a new option flag for the ElasticsearchWriter to toggle the new format. Then it would not be a breaking change.

**Comment by julianbrost** [comment]
Our strategy is to focus on the `OTLPMetricsWriter` (#10685) and deprecate the existing `ElasticsearchWriter` in favor it (#10730). That also happens to answer the question of backwards-compatibility :slightly_smiling_face:.

Can someone please confirm that the issue described here was addressed or isn't an issue in the OTLP writer? Then I'd close this PR.

**Comment by martialblog** [comment]
Hi, I did test the OTLPMetricsWriter with Elasticsearch and OpenSearch. Both would work for my use cases. Currently OpenSearch does not have an API endpoint for OTLP, so you need a tool in between Icinga2 and OpenSearch, but the OpenSearch DataPrepper works the OTLPMetricsWriter.

The issue described here is addressed. The OTLP endpoint in Elasticsearch handles creation on a datastream and a field mapping template. So there should be no mapping explosion.

I have details on that here: https://github.com/Icinga/icinga2/pull/10685#issuecomment-3854036469

**Comment by julianbrost** [comment]
> so you need a tool in between Icinga2 and OpenSearch, but the OpenSearch DataPrepper works the OTLPMetricsWriter

Though that's a tool maintained by OpenSearch themselves. Yes, it's an extra step, but it's not like it would require some sketchy tool. And if in doubt, I'd rather spend the time on providing some "setting up Icinga 2 + OpenSearch + Data Prepper" documentation rather than maintaining a whole writer implementation long-term.

Nonetheless, thanks for bringing this up and providing all kinds of feedback for our writers. Even with this PR closed, I think we'll make good progress with our writers in v2.16 thanks to your help. :)
